minimal_data_sharing/minimal_atomic_parameter.cpp

```diff
@@ -0,0 +1,94 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+using namespace std::chrono_literals;
+
+static_assert(std::atomic<std::uint32_t>::is_always_lock_free);
+
+class MinimalPublisher : public rclcpp::Node
+{
+public:
+  MinimalPublisher()
+  : Node("minimal_publisher"), count_(0)
+  {
+    publisher_ = this->create_publisher<std_msgs::msg::String>("topic", 10);
+    auto timer_callback =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello, world! " + std::to_string(this->count_++);
+        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
+        this->publisher_->publish(message);
+      };
+
+    realtime_callback_group_ = this->create_callback_group(
+      rclcpp::CallbackGroupType::MutuallyExclusive, false);
+    timer_ = this->create_wall_timer(500ms, timer_callback, realtime_callback_group_);
+
+    count_ = declare_parameter("count", 0);
+    parameter_event_handler_ = std::make_shared<rclcpp::ParameterEventHandler>(this);
+    auto param_callback = [this](const rclcpp::Parameter & p) {
+        RCLCPP_INFO(this->get_logger(), "Count changed to: '%ld'", p.as_int());
+        count_.store(static_cast<std::uint32_t>(p.as_int()));
+      };
+    parameter_callback_handle_ =
+      parameter_event_handler_->add_parameter_callback("count", param_callback);
+  }
+
+  rclcpp::CallbackGroup::SharedPtr get_realtime_callback_group()
+  {
+    return realtime_callback_group_;
+  }
+
+private:
+  rclcpp::TimerBase::SharedPtr timer_;
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
+  std::atomic<std::uint32_t> count_;
+  rclcpp::CallbackGroup::SharedPtr realtime_callback_group_;
+  std::shared_ptr<rclcpp::ParameterEventHandler> parameter_event_handler_;
+  std::shared_ptr<rclcpp::ParameterCallbackHandle> parameter_callback_handle_;
+};
+
+
+int main(int argc, char * argv[])
+{
+  rclcpp::init(argc, argv);
+
+  auto node = std::make_shared<MinimalPublisher>();
+  rclcpp::executors::StaticSingleThreadedExecutor default_callback_group_executor;
+  rclcpp::executors::StaticSingleThreadedExecutor realtime_executor;
+
+  default_callback_group_executor.add_node(node);
+  realtime_executor.add_callback_group(
+    node->get_realtime_callback_group(), node->get_node_base_interface());
+
+  // spin real-time tasks in a separate thread
+  auto realtime_thread = std::thread(
+    [&]() {
+      realtime_executor.spin();
+    });
+
+  default_callback_group_executor.spin();
+  realtime_thread.join();
```

Comment:
I may be wrong, but this thread will never join as it spins forever right? I think Ctrl+C or similar will not properly shutdown the application by calling `rclcpp::shutdown`. How is the shutdown supposed to work?

Reply:
It should work. The default signal handler will shut down the context when Ctrl+C is used (see [utilities.hpp](https://github.com/ros2/rclcpp/blob/03fa731d2345c41e5a4e436bf0730012ece96ce5/rclcpp/include/rclcpp/utilities.hpp#L87-L89). Then, the spin will unblock when the context is shut down (see [executor.cpp](https://github.com/ros2/rclcpp/blob/03fa731d2345c41e5a4e436bf0730012ece96ce5/rclcpp/src/rclcpp/executor.cpp#L444)).